# Python基础3—上下文管理器

<!-- TOC -->

- [Python基础3—上下文管理器](#python基础3上下文管理器)
    - [1. 术语](#1-术语)
    - [2. 基本语法和工作原理](#2-基本语法和工作原理)
        - [2.1 语法格式](#21-语法格式)
        - [2.2 基本例子](#22-基本例子)
        - [2.3 with语句内部执行过程](#23-with语句内部执行过程)
    - [3. contextlib模块](#3-contextlib模块)

<!-- /TOC -->

## 1. 术语

- **上下文管理器协议** (Context Management Protocol): 指要实现对象的`__enter__()`和`__exit__()`方法
- **上下文管理器** (Context Manager)：支持上下文管理器协议的对象，也就是实现了`__enter__()`和`__exit__()`方法
- **语句体** (with-body): with语句包裹起来的代码块
- **上下文表达式** (Context Expression): with语句中紧跟在with之后的表达式
- **运行时上下文** (Runtime Context): `__enter__()`和`__exit_()`方法

## 2. 基本语法和工作原理

### 2.1 语法格式

```python
with context_expression [as target(s)]:
    with-body
```

这里context_expression返回一个上下文管理器对象。如果指定了as语句，则将`__enter__()`方法的返回值赋值给target(s)。target(s)可是单个变量，或者是由()括起来的元组。

### 2.2 基本例子

```python
class Function(object):
    def __init__(self, x, y):
        print 'initialize x and y'
        self.x, self.y = x, y
    def __enter__(self):
        print '[Entering context]'
        return self  # 也可以返回其他值
   	def __exit__(self, type, value, traceback):  # 或者写成(self, *args)
        print '[Exiting context]'
    def get_sum(self):
        sum_ = sum(x, y)
        return sum
```

用with语句调用：

```python
with Function(1, 2) as f:
    print 'sum:', f.get_sum()
```

输出结果：

```shell
initialize x and y  # 调用 __init__
[Entering context]  # 调用 __enter__
sum: 3              # 调用 with-body
[Exiting context]   # 调用 __exit__
```

代码分析：

- Function(1, 2) 生成了一个上下文管理器
- 调用`__enter__()`方法，并将返回值给with后的f
- 执行语句体with-body
- 执行`__exit__()`方法

说明：

(1) 不管执行语句体的过程是否发生异常，都会执行上下文管理器的`__exit__()`方法。因此，`__exit__()`方法一般用于负责“清理”工作，如释放资源，关闭文件等。

(2) 如果with-body执行过程没有发生异常，或者with-body中出现了break/continue/return，则会以None参数调用`__exit__(None, None, None)`；如果with-body执行时出现了异常，则会调用`sys.exc_info`获取异常信息，用得到的三个参数调用`__exit__(exc_type, exc_value, exc_traceback)`

(3) 发生异常时，调用`__exit__(exc_type, exc_value, exc_traceback)`。该函数默认会返回None值，而布尔环境中None值就是False。在这种情况下(返回值为False)，抛出异常，让with外面的语句自动去处理异常。如果显示强行返回True，则会忽略异常。(如果上面例子的`__exit__()`语句最后一行添加一行`return True`，则无论怎样也不会报异常了)

(4) 若with-body出现异常触发`__exit__(exc_type, exc_value, exc_traceback)`时该语句又出现了异常，则`__exit__(exc_type, exc_value, exc_traceback)`中的异常会取代with-body中的异常。

(5) 注意在自定义上下文管理器时，要处理异常时，不能在`__exit__()`语句中重新抛出异常，只需将返回值设为False，或者什么也不设置(默认放回None，None也是False)，然后由外面的代码来自动处理异常。

### 2.3 with语句内部执行过程

明白了以上几点，我们再来看with语句内部具体是如何实现的：

```python
context_manager = context_expression
    exit = type(context_manager).__exit__  
    value = type(context_manager).__enter__(context_manager)
    exc = True   # True 表示正常执行，即便有异常也忽略；False 表示重新抛出异常，需要对异常进行处理
    try:
        try:
            target = value  # 如果使用了 as 子句
            with-body     # 执行 with-body
        except:
            # 执行过程中有异常发生
            exc = False
            # 如果 __exit__ 返回 True，则异常被忽略；如果返回 False，则重新抛出异常
            # 由外层代码对异常进行处理
            if not exit(context_manager, *sys.exc_info()):
                raise
    finally:
        # 正常退出，或者通过 statement-body 中的 break/continue/return 语句退出
        # 或者忽略异常退出
        if exc:
            exit(context_manager, None, None, None) 
        # 缺省返回 None，None 在布尔上下文中看做是 False
```

可以尝试自定义with语句:

```python
class DummyResource(object):
    def __init__(self, tag):
        self.tag = tag
        print 'Resource [%s]' % tag
    def __enter__(self):
        print '[Enter %s]: Allocate resource.' % self.tag
        return self	  # 可以返回不同的对象
    def __exit__(self, exc_type, exc_value, exc_tb):
        print '[Exit %s]: Free resource.' % self.tag  # 这里释放资源
        if exc_tb is None:
            print '[Exit %s]: Exited without exception.' % self.tag
        else:
            print '[Exit %s]: Exited with exception raised.' % self.tag
            return False   # 可以省略，缺省的None也是被看做是False
```

## 3. contextlib模块

contextlib 模块提供了3个对象：装饰器 contextmanager、函数 nested 和上下文管理器 closing。使用这些对象，可以对已有的生成器函数或者对象进行包装，加入对上下文管理协议的支持，避免了专门编写上下文管理器来支持 with 语句。

这里主要写一下contextmanager的用法：

contextmanager用于对生成器函数进行装饰，生成器被装饰后返回的是一个上下文管理器，其`__enter__()`和`__exit__()`方法由contextmanager负责提供。也就是说，yeild之前的内容为`__enter__()`，之后的内容为`__exit()__`。

另外，被装饰的生成器函数只能产生一个值，否则会导致RuntimeError异常。如果使用了as语句，则产生的值赋值给as语句中的target:

```python
from contextlib import contextmanager

@contextmanager
def demo():
    print '[Allocate resources]'
    print 'Code before yield-statement executes in __enter__'
    yield '*** contextmanager demo ***'
    print 'Code after yield-statement executes in __exit__'
    print '[Free resources]'

with demo() as value:
    print 'Assigned Value: %s' % value
```

输出结果如下:

```shell
[Allocate resources]
Code before yield-statement executes in __enter__
Assigned Value: *** contextmanager demo ***
Code after yield-statement executes in __exit__
[Free resources]
```

---

参考链接:

[浅谈Python的with语句](https://www.ibm.com/developerworks/cn/opensource/os-cn-pythonwith/#icomments)

[上下文管理器·Python之旅](http://funhacks.net/explore-python/Advanced-Features/context.html)



